sycl/include/CL/sycl/ONEAPI/sub_group.hpp

```diff
@@ -295,29 +295,70 @@ struct sub_group {
                         PI_INVALID_DEVICE);
 #endif
   }
-
+#ifdef __SYCL_DEVICE_ONLY__
+#ifdef __NVPTX__
   template <int N, typename T, access::address_space Space>
   sycl::detail::enable_if_t<
-      sycl::detail::sub_group::AcceptableForGlobalLoadStore<T, Space>::value &&
-          N != 1,
+      sycl::detail::sub_group::AcceptableForGlobalLoadStore<T, Space>::value,
       vec<T, N>>
   load(const multi_ptr<T, Space> src) const {
-#ifdef __SYCL_DEVICE_ONLY__
-#ifdef __NVPTX__
     vec<T, N> res;
     for (int i = 0; i < N; ++i) {
       res[i] = *(src.get() + i * get_max_local_range()[0] + get_local_id()[0]);
     }
     return res;
-#else
+  }
+#else  // __NVPTX__
+  template <int N, typename T, access::address_space Space>
+  sycl::detail::enable_if_t<
+      sycl::detail::sub_group::AcceptableForGlobalLoadStore<T, Space>::value &&
+          N != 1 && N != 3 && N != 16,
+      vec<T, N>>
+  load(const multi_ptr<T, Space> src) const {
     return sycl::detail::sub_group::load<N, T>(src);
-#endif // __NVPTX__
-#else
+  }
+
+  template <int N, typename T, access::address_space Space>
+  sycl::detail::enable_if_t<
+      sycl::detail::sub_group::AcceptableForGlobalLoadStore<T, Space>::value &&
+          N == 16,
+      vec<T, 16>>
+  load(const multi_ptr<T, Space> src) const {
+    return {sycl::detail::sub_group::load<8, T>(src),
+            sycl::detail::sub_group::load<8, T>(src +
+                                                8 * get_max_local_range()[0])};
+  }
+
+  template <int N, typename T, access::address_space Space>
+  sycl::detail::enable_if_t<
+      sycl::detail::sub_group::AcceptableForGlobalLoadStore<T, Space>::value &&
+          N == 3,
+      vec<T, 3>>
+  load(const multi_ptr<T, Space> src) const {
+    auto res = sycl::detail::sub_group::load<4, T>(src);
```

Comment:
This is fine for `vec` (because SYCL says that `vec<4>` has the alignment requirements of `vec<3>`), but we probably need to revisit this when we change the return type to `marray<3>`.  I don't think a user would expect a sub-group load of 3 elements to potentially read off the end of an array, which some implementations of `load<4>` might do.

I think we should add a TODO here, or split into `load<1>` and `load<2>` as you did for `store`.